### DISCUSSION TITLE
Botan 3 migration question (Argon2)

### DISCUSSION THREAD

**Discussion by Tectu** [randombit/botan#4033] [Q&A] [answered]
I'm currently migrating a code base from Botan 2 to Botan 3.
Part of the code base uses Argon2 for password hashing:

Create hash (simplified):
```
std::string pw;
std::size_t passes;
std::size_t memory_size;
std::size_t variant; // 0 = Argon2d,  1 = Argon2i,  2 = Argon 2id
std::size_t salt_length;
std::size_t output_length;

// ...
    
std::string hash = Botan::argon2_generate_pwhash(
    pw.data(),
    pw.size(),
    m_rng,
    1,              // p (currently Botan does not support anything other than 1)
    memory_size,    // M
    passes,         // t
    variant,        // y (mode)
    salt_length,
    output_length
);
```
Check hash (simplified):
```
std::size_t hash;

// ...
const bool matches = Botan::argon2_check_pwhash(
    pw.data(),
    pw.size(),
    hash
);
```

Migrating to Botan 3, I'm supposed to use `Botan::PasswordHashFamily::create()`.
That `create()` function is expecting an `algo_spec` (which is presumably "Argon2d", "Argon2i", etc).

Right now I have two questions:

1. When checking an existing password hash (previously using `Botan::argon2_check_pwhash()`), I also need to use the `Botan::PasswordHashFamily::create()` factory. However, I'm not sure what to pass as the `algo_spec` in that case. When checking an existing Argon2 password hash, we only know that it's an Argon2 hash, but we don't know the variant. How is this supposed to be handled?
2. Previously, we were able to explicitly specify `p`, `M` and `t` parameters of the Argon2 generator algorithm using `Botan::argon2_generate_pwhash()`. How does this work with the new approach? There's `PasswordHashFamily::tune()` but that does (understandably) not seem to provide algorithm specific tunables.

**Comment by randombit** [comment, marked as answer]
Oh sorry about that, this should have been in the migration guide - `argon2_generate_pwhash` and `argon2_check_pwhash` are still exactly as they were, but the declarations moved to `argon2fmt.h`. Also (in case you are doing something with module selection, etc in your build), these functions are implemented in a new module `argon2fmt` (whereas before they were bundled with everything else in `argon2`)

Given this I don't think your other questions are relevant to your problem but for future readers/etc:

1. You would do it exactly as [argon2_check_pwhash](https://github.com/randombit/botan/blob/702d26d88db3e19a6f68843d05929f1508f09734/src/lib/passhash/argon2fmt/argon2fmt.cpp#L86-L94) does, by looking at the prefix of the hash.

2. There is also `PasswordHashFamily::from_params` which takes three parameters, which for Argon2 are interpreted as `M`, `t`, and `p` in that order.

**Reply by randombit** [reply]
BTW I just noticed the comment "p (currently Botan does not support anything other than 1)"

Argon2 `p > 1` was always supported in that you could set a higher `p`, but in the first releases it did not take advantage of the available parallelism. Since 3.0.0 and 2.19.2, Argon2 with `p > 1` is accelerated on systems with thread support.

**Reply by Tectu** [reply]
Thank you - that worked well!

Also thank you for the additional information regarding the `p` parameter. This code was indeed written quite some time ago - I should have just omitted the comment from the simplified code example but at least now we know :D

> Also (in case you are doing something with module selection, etc in your build),

That was also helpful too :)
I'm indeed doing that. A while ago I created a `FindBotan.cmake` script which allows to use Botan with CMake's `find_package()` but using the amalgamation build mechanism of Botan instead of manually dealing with maintaining file references.
It's not perfect, but it is a lot easier to deal with when targeting a lot of different platforms.
In case anybody is interested: https://github.com/Tectu/botan-cmake
